File: slr-ide/python_engine/llm/vault.py
import base64
import os
import bcrypt
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes

# Configuration Constants
PBKDF2_ITERATIONS = 600000

def hash_master_password(password: str) -> str:
    """Hashes the master password using bcrypt."""
    salt = bcrypt.gensalt()
    hashed = bcrypt.hashpw(password.encode('utf-8'), salt)
    return hashed.decode('utf-8')

def verify_master_password(password: str, hashed_password: str) -> bool:
    """Verifies a master password against its bcrypt hash."""
    try:
        return bcrypt.checkpw(password.encode('utf-8'), hashed_password.encode('utf-8'))
    except Exception:
        return False

def derive_key(password: str, salt: bytes) -> bytes:
    """Derives a 256-bit AES key from password and salt using PBKDF2."""
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=PBKDF2_ITERATIONS
    )
    return kdf.derive(password.encode('utf-8'))

def encrypt_key(plain_key: str, password: str) -> dict:
    """Encrypts a plaintext key using AES-256-GCM derived from the password.
    Returns base64 encoded ciphertext, salt, iv, and tag.
    """
    salt = os.urandom(32)
    iv = os.urandom(12)
    derived_aes_key = derive_key(password, salt)
    
    aesgcm = AESGCM(derived_aes_key)
    # The encrypt method in cryptography's AESGCM combines ciphertext and tag
    ciphertext_and_tag = aesgcm.encrypt(iv, plain_key.encode('utf-8'), None)
    
    # Separate ciphertext and tag (tag is the last 16 bytes in cryptography AESGCM)
    ciphertext = ciphertext_and_tag[:-16]
    tag = ciphertext_and_tag[-16:]
    
    return {
        "ciphertext": base64.b64encode(ciphertext).decode('utf-8'),
        "salt": base64.b64encode(salt).decode('utf-8'),
        "iv": base64.b64encode(iv).decode('utf-8'),
        "tag": base64.b64encode(tag).decode('utf-8')
    }

def decrypt_key(encrypted_data: dict, password: str) -> str:
    """Decrypts AES-256-GCM data using the password."""
    ciphertext = base64.b64decode(encrypted_data["ciphertext"])
    salt = base64.b64decode(encrypted_data["salt"])
    iv = base64.b64decode(encrypted_data["iv"])
    tag = base64.b64decode(encrypted_data["tag"])
    
    derived_aes_key = derive_key(password, salt)
    aesgcm = AESGCM(derived_aes_key)
    
    # Combine ciphertext and tag for decryption
    ciphertext_and_tag = ciphertext + tag
    decrypted_bytes = aesgcm.decrypt(iv, ciphertext_and_tag, None)
    
    return decrypted_bytes.decode('utf-8')

File: slr-ide/src/lib/vault.ts
import crypto from 'crypto';

const PBKDF2_ITERATIONS = 600000;

/**
 * Derives a 256-bit AES key from a password and salt using PBKDF2.
 */
export function deriveKey(password: string, salt: Buffer): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    crypto.pbkdf2(password, salt, PBKDF2_ITERATIONS, 32, 'sha256', (err, derivedKey) => {
      if (err) reject(err);
      else resolve(derivedKey);
    });
  });
}

interface EncryptedData {
  ciphertext: string;
  salt: string;
  iv: string;
  tag: string;
}

/**
 * Encrypts a key using AES-256-GCM.
 */
export async function encryptKey(plainKey: string, password: string): Promise<EncryptedData> {
  const salt = crypto.randomBytes(32);
  const iv = crypto.randomBytes(12);
  const key = await deriveKey(password, salt);

  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
  let ciphertext = cipher.update(plainKey, 'utf8', 'base64');
  ciphertext += cipher.final('base64');
  const tag = cipher.getAuthTag();

  return {
    ciphertext,
    salt: salt.toString('base64'),
    iv: iv.toString('base64'),
    tag: tag.toString('base64'),
  };
}

/**
 * Decrypts a key using AES-256-GCM.
 */
export async function decryptKey(encrypted: EncryptedData, password: string): Promise<string> {
  const salt = Buffer.from(encrypted.salt, 'base64');
  const iv = Buffer.from(encrypted.iv, 'base64');
  const tag = Buffer.from(encrypted.tag, 'base64');
  const key = await deriveKey(password, salt);

  const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv);
  decipher.setAuthTag(tag);
  let decrypted = decipher.update(encrypted.ciphertext, 'base64', 'utf8');
  decrypted += decipher.final('utf8');

  return decrypted;
}
